### PR TITLE
feat: add etag for objectMeta

### LINKED ISSUE #2240: Add ETag to ObjectMeta
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

All object stores support a notion of an ETag that uniquely identifies a resource, we should support this

**Describe the solution you'd like**

The challenge lies in how to support this for the InMemory and LocalFileSystem stores. In lieu of a better alternative, as described [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) a hash of the last modification time is likely sufficient.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Weijun-H**
# Which issue does this PR close?

Closes #2240

# Rationale for this change
Add `Etag` for `objectMeta` as a unique identifier, which is a hash value by `last_modified`.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
For the actual stores, like S3, the etag is part of the response. Using a manual hash is just for in memory and filesystem where this data isn't stored

**Comment by tustvold** [comment]
The emulator tests appear to be failing for this?

**Comment by Weijun-H** [comment]
> The emulator tests appear to be failing for this?

```
---- http::tests::http_test stdout ----
thread 'http::tests::http_test' panicked at 'called `Result::unwrap()` on an `Err` value: Generic { store: "HTTP", source: InvalidPropFind { source: Custom("missing field `getetag`") } }', object_store/src/lib.rs:1203:62
```

`getetag` is meant to work, but it failed.

```
[13.6](https://datatracker.ietf.org/doc/html/rfc2518#section-13.6) getetag Property

   Name:       getetag
   Namespace:  DAV:
   Purpose:    Contains the ETag header returned by a GET without
   accept headers.
   Description: The getetag property MUST be defined on any DAV
   compliant resource that returns the Etag header.
   Value:      entity-tag  ; defined in [section 3.11 of [RFC2068]](https://datatracker.ietf.org/doc/html/rfc2068#section-3.11)

   <!ELEMENT getetag (#PCDATA) >
```

**Comment by tustvold** [comment]
It may not be supported, we may need to make the ETag optional 🤔

File: object_store/src/lib.rs [hunk 1]
```diff
@@ -518,6 +518,8 @@ pub struct ObjectMeta {
     pub last_modified: DateTime<Utc>,
     /// The size in bytes of the object
     pub size: usize,
+    /// The unique identifier for the object
+    pub e_tag: String,
```
Comment (tustvold) [inline object_store/src/lib.rs:522]:
```suggestion
    pub e_tag: Option<String>,
```
Perhaps?

Comment (tustvold) [reply object_store/src/lib.rs:522]: In fact returning `None` might be better than using a hash of datetime for the stores such as LocalFilesystem. Makes it explicit. What do you think?

Comment (Weijun-H) [reply object_store/src/lib.rs:522]: Make it to `Option` is better. But I think I still need to assign a  manual `Etag` for LocalFilesystem.

Comment (tustvold) [reply object_store/src/lib.rs:522]: Why not just return None? If a downstream wants a hash of the last modification time they can always compute this themselves?

Comment (Weijun-H) [reply object_store/src/lib.rs:522]: Yes, you are right. Return None is more reasonable

File: object_store/Cargo.toml [hunk 2]
```diff
@@ -58,9 +58,11 @@ aws-config = { version = "0.54", optional = true }
 
 [target.'cfg(not(target_arch = "wasm32"))'.dependencies]
 tokio = { version = "1.25.0", features = ["sync", "macros", "rt", "time", "io-util", "fs"] }
+ahash = { version = "0.8", default-features = false, features = ["runtime-rng"] }
```
Comment (tustvold) [inline object_store/Cargo.toml:61]: Do we still need this?

**Review by tustvold — APPROVED** [review]
Looks good to me, thank you, as this is a breaking change I'm going to hold off on merging this until I've cut the next point release which I intend to do this wewk

File: object_store/Cargo.toml (hunk 2 above)
Comment (Weijun-H) [reply object_store/Cargo.toml]: No, we don't.

**Pull request merged**